// @kolkov/ngx-gallery/esm5/lib/ngx-gallery-bullets/ngx-gallery-bullets.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var NgxGalleryBulletsComponent = /** @class */ (function () {
    function NgxGalleryBulletsComponent() {
        this.active = 0;
        // tslint:disable-next-line:no-output-native
        this.change = new EventEmitter();
    }
    NgxGalleryBulletsComponent.prototype.ngOnInit = function () {
    };
    NgxGalleryBulletsComponent.prototype.getBullets = function () {
        return Array(this.count);
    };
    NgxGalleryBulletsComponent.prototype.handleChange = function (event, index) {
        this.change.emit(index);
    };
    tslib_1.__decorate([
        Input()
    ], NgxGalleryBulletsComponent.prototype, "count", void 0);
    tslib_1.__decorate([
        Input()
    ], NgxGalleryBulletsComponent.prototype, "active", void 0);
    tslib_1.__decorate([
        Output()
    ], NgxGalleryBulletsComponent.prototype, "change", void 0);
    NgxGalleryBulletsComponent = tslib_1.__decorate([
        Component({
            selector: 'ngx-gallery-bullets',
            template: "<div class=\"ngx-gallery-bullet\" *ngFor=\"let bullet of getBullets(); let i = index;\" (click)=\"handleChange($event, i)\"\r\n     [ngClass]=\"{ 'ngx-gallery-active': i == active }\"></div>\r\n",
            styles: [":host{position:absolute;z-index:2000;display:-webkit-inline-box;display:inline-flex;left:50%;-webkit-transform:translateX(-50%);transform:translateX(-50%);bottom:0;padding:10px}.ngx-gallery-bullet{width:10px;height:10px;border-radius:50%;cursor:pointer;background:#fff}.ngx-gallery-bullet:not(:first-child){margin-left:5px}.ngx-gallery-bullet.ngx-gallery-active,.ngx-gallery-bullet:hover{background:#000}"]
        })
    ], NgxGalleryBulletsComponent);
    return NgxGalleryBulletsComponent;
}());
export { NgxGalleryBulletsComponent };
