// @kolkov/ngx-gallery/esm2015/lib/ngx-gallery.module.js
import * as tslib_1 from "tslib";
import { NgModule } from '@angular/core';
import { NgxGalleryComponent } from './ngx-gallery.component';
import { NgxGalleryImageComponent } from './ngx-gallery-image/ngx-gallery-image.component';
import { NgxGalleryArrowsComponent } from './ngx-gallery-arrows/ngx-gallery-arrows.component';
import { NgxGalleryThumbnailsComponent } from './ngx-gallery-thumbnails/ngx-gallery-thumbnails.component';
import { NgxGalleryPreviewComponent } from './ngx-gallery-preview/ngx-gallery-preview.component';
import { NgxGalleryActionComponent } from './ngx-gallery-action/ngx-gallery-action.component';
import { NgxGalleryBulletsComponent } from './ngx-gallery-bullets/ngx-gallery-bullets.component';
import { CommonModule } from '@angular/common';
import { HAMMER_GESTURE_CONFIG, HammerGestureConfig } from '@angular/platform-browser';
export class CustomHammerConfig extends HammerGestureConfig {
    constructor() {
        super(...arguments);
        this.overrides = {
            pinch: { enable: false },
            rotate: { enable: false }
        };
    }
}
let NgxGalleryModule = class NgxGalleryModule {
};
NgxGalleryModule = tslib_1.__decorate([
    NgModule({
        declarations: [
            NgxGalleryComponent,
            NgxGalleryImageComponent,
            NgxGalleryArrowsComponent,
            NgxGalleryThumbnailsComponent,
            NgxGalleryPreviewComponent,
            NgxGalleryActionComponent,
            NgxGalleryBulletsComponent
        ],
        imports: [CommonModule],
        exports: [NgxGalleryComponent],
        providers: [
            { provide: HAMMER_GESTURE_CONFIG, useClass: CustomHammerConfig }
        ]
    })
], NgxGalleryModule);
export { NgxGalleryModule };
