// @kolkov/ngx-gallery/esm5/lib/ngx-gallery-arrows/ngx-gallery-arrows.component.js
import * as tslib_1 from "tslib";
import { Component, EventEmitter, Input, Output } from '@angular/core';
var NgxGalleryArrowsComponent = /** @class */ (function () {
    function NgxGalleryArrowsComponent() {
        this.prevClick = new EventEmitter();
        this.nextClick = new EventEmitter();
    }
    NgxGalleryArrowsComponent.prototype.ngOnInit = function () {
    };
    NgxGalleryArrowsComponent.prototype.handlePrevClick = function () {
        this.prevClick.emit();
    };
    NgxGalleryArrowsComponent.prototype.handleNextClick = function () {
        this.nextClick.emit();
    };
    tslib_1.__decorate([
        Input()
    ], NgxGalleryArrowsComponent.prototype, "prevDisabled", void 0);
    tslib_1.__decorate([
        Input()
    ], NgxGalleryArrowsComponent.prototype, "nextDisabled", void 0);
    tslib_1.__decorate([
        Input()
    ], NgxGalleryArrowsComponent.prototype, "arrowPrevIcon", void 0);
    tslib_1.__decorate([
        Input()
    ], NgxGalleryArrowsComponent.prototype, "arrowNextIcon", void 0);
    tslib_1.__decorate([
        Output()
    ], NgxGalleryArrowsComponent.prototype, "prevClick", void 0);
    tslib_1.__decorate([
        Output()
    ], NgxGalleryArrowsComponent.prototype, "nextClick", void 0);
    NgxGalleryArrowsComponent = tslib_1.__decorate([
        Component({
            selector: 'ngx-gallery-arrows',
            template: "<div class=\"ngx-gallery-arrows-wrapper ngx-gallery-arrow-left\">\r\n  <div class=\"ngx-gallery-icon ngx-gallery-arrow\"  (click)=\"handlePrevClick()\" [class.ngx-gallery-disabled]=\"prevDisabled\">\r\n    <i class=\"ngx-gallery-icon-content fa fa-arrow-circle-left\"></i>\r\n  </div>\r\n</div>\r\n<div class=\"ngx-gallery-arrows-wrapper ngx-gallery-arrow-right\">\r\n  <div class=\"ngx-gallery-icon ngx-gallery-arrow\"  (click)=\"handleNextClick()\" [class.ngx-gallery-disabled]=\"nextDisabled\">\r\n    <i class=\"ngx-gallery-icon-content fa fa-arrow-circle-right\"></i>\r\n  </div>\r\n</div>\r\n",
            styles: [".ngx-gallery-arrow-wrapper{position:absolute;height:100%;width:1px;display:table;z-index:2000;table-layout:fixed}.ngx-gallery-arrow-left{left:0}.ngx-gallery-arrow-right{right:0}.ngx-gallery-arrow{top:50%;-webkit-transform:translateY(-50%);transform:translateY(-50%);cursor:pointer}.ngx-gallery-arrow.ngx-gallery-disabled{opacity:.6;cursor:default}.ngx-gallery-arrow-left .ngx-gallery-arrow{left:10px}.ngx-gallery-arrow-right .ngx-gallery-arrow{right:10px}"]
        })
    ], NgxGalleryArrowsComponent);
    return NgxGalleryArrowsComponent;
}());
export { NgxGalleryArrowsComponent };
