// @kolkov/ngx-gallery/esm5/lib/ngx-gallery-action/ngx-gallery-action.component.js
import * as tslib_1 from "tslib";
import { ChangeDetectionStrategy, Component, EventEmitter, Input, Output } from '@angular/core';
var NgxGalleryActionComponent = /** @class */ (function () {
    function NgxGalleryActionComponent() {
        this.disabled = false;
        this.titleText = '';
        // tslint:disable-next-line:no-output-native
        this.click = new EventEmitter();
    }
    NgxGalleryActionComponent.prototype.ngOnInit = function () {
    };
    NgxGalleryActionComponent.prototype.handleClick = function (event) {
        if (!this.disabled) {
            this.click.emit(event);
        }
        event.stopPropagation();
        event.preventDefault();
    };
    tslib_1.__decorate([
        Input()
    ], NgxGalleryActionComponent.prototype, "icon", void 0);
    tslib_1.__decorate([
        Input()
    ], NgxGalleryActionComponent.prototype, "disabled", void 0);
    tslib_1.__decorate([
        Input()
    ], NgxGalleryActionComponent.prototype, "titleText", void 0);
    tslib_1.__decorate([
        Output()
    ], NgxGalleryActionComponent.prototype, "click", void 0);
    NgxGalleryActionComponent = tslib_1.__decorate([
        Component({
            selector: 'ngx-gallery-action',
            template: "<div class=\"ngx-gallery-icon\" [class.ngx-gallery-icon-disabled]=\"disabled\"\r\n     aria-hidden=\"true\"\r\n     title=\"{{ titleText }}\"\r\n     (click)=\"handleClick($event)\">\r\n  <i class=\"ngx-gallery-icon-content {{ icon }}\"></i>\r\n</div>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            styles: [".ngx-gallery-icon{color:#fff;z-index:2000;display:inline-block;position:relative;margin-right:10px;margin-top:10px;font-size:25px;cursor:pointer;text-decoration:none}.ngx-gallery-icon .ngx-gallery-icon-content{display:block}"]
        })
    ], NgxGalleryActionComponent);
    return NgxGalleryActionComponent;
}());
export { NgxGalleryActionComponent };
